crates/sui-framework/packages/dwallet-system/sources/dwallet.move

```diff
@@ -0,0 +1,265 @@
+        let holder = MessageApprovalsHolder {
+            id: object::new(ctx),
+            message_approvals,
+        };
+        transfer::transfer(holder, tx_context::sender(ctx));
+    }
+
+    public fun remove_message_approvals_holder(holder: MessageApprovalsHolder): vector<MessageApproval> {
+        let MessageApprovalsHolder {
+            id,
+            message_approvals,
+        } = holder;
+        object::delete(id);
+        message_approvals
+    }
+
+    public fun approve_messages(dwallet_cap: &DWalletCap, messages: vector<vector<u8>>): vector<MessageApproval> {
+        let dwallet_cap_id = object::id(dwallet_cap);
+        let message_approvals = vector::empty<MessageApproval>();
+        while (vector::length(&messages) > 0) {
+            let message = vector::pop_back(&mut messages);
+            vector::push_back(&mut message_approvals, MessageApproval {
+                dwallet_cap_id,
+                message,
+            });
+        };
+        message_approvals
+    }
+
+    public fun message_approval_dwallet_cap_id(message_approval: &MessageApproval): ID {
+        message_approval.dwallet_cap_id
+    }
+
+    public fun message_approval_message(message_approval: &MessageApproval): vector<u8> {
+        message_approval.message
+    }
+
+    public fun remove_message_approval(message_approval: MessageApproval): (ID, vector<u8>) {
+        let MessageApproval {
+            dwallet_cap_id,
+            message
+        } = message_approval;
+        (dwallet_cap_id, message)
+    }
+
+    public fun create_shared_partial_user_signed_messages<S: store, E: store + copy + drop>(
+        partial_user_signed_messages: PartialUserSignedMessages<S, E>,
+        ctx: &mut TxContext
+    ) {
+        let holder = SharedPartialUserSignedMessages {
+            id: object::new(ctx),
+            partial_user_signed_messages,
+        };
+        transfer::share_object(holder);
+    }
+
+    public fun sign_shared<S: store, E: store + copy + drop>(
+        shared: SharedPartialUserSignedMessages<S, E>,
+        message_approvals: vector<MessageApproval>,
+        ctx: &mut TxContext
+    ) {
+        let SharedPartialUserSignedMessages {
+            id,
+            partial_user_signed_messages,
+        } = shared;
+        object::delete(id);
+        sign(partial_user_signed_messages, message_approvals, ctx)
+    }
+
+    public(friend) fun create_partial_user_signed_messages<S: store, E: store + copy + drop>(
+        dwallet_id: ID,
+        dwallet_cap_id: ID,
+        messages: vector<vector<u8>>,
+        sign_data: S,
+        sign_data_event: E,
+        ctx: &mut TxContext
+    ): PartialUserSignedMessages<S, E> {
+        PartialUserSignedMessages {
+            id: object::new(ctx),
+            dwallet_id,
+            dwallet_cap_id,
+            messages,
+            sign_data,
+            sign_data_event,
+        }
```

Comment:
@omersadika can we take the dwallet_cap by reference here instead of the ID?
I know that the caller is a friend and will perform security validations there, but it is safer to have it anyway if possible.

Reply:
Or is the caller of this function not necessairly an owner of the cap?

Reply:
Cant